src/coreclr/jit/assertionprop.cpp

```diff
@@ -5053,6 +5054,75 @@ static GCInfo::WriteBarrierForm GetWriteBarrierForm(ValueNumStore* vnStore, Valu
             // Boxed static - always on the heap
             return GCInfo::WriteBarrierForm::WBF_BarrierUnchecked;
         }
+        if (funcApp.m_func == VNF_InitVal)
+        {
+            // Check if we deal with `ref ByrefLikeStruct` argument
+            unsigned lclNum = (unsigned)vnStore->CoercedConstantValue<ssize_t>(funcApp.m_args[0]);
+            assert(lclNum != BAD_VAR_NUM);
+            if (comp->lvaIsParameter(lclNum))
+            {
+                if (comp->info.compThisArg == lclNum)
+                {
+                    // No barrier needed if current class is byref-like
+                    // and the destination is an implicit "this"
+                    return comp->eeIsByrefLike(comp->info.compClassHnd) ? GCInfo::WriteBarrierForm::WBF_NoBarrier
+                                                                        : GCInfo::WriteBarrierForm::WBF_BarrierUnknown;
+                }
+
+                // Find the corresponding argument index. Basically, we need to ignore the hidden args
+                // such as generic context, return buffer, etc..
+                unsigned ilArg = comp->compMap2ILvarNum(lclNum);
+                switch (static_cast<int>(ilArg))
+                {
+                    // Ignore non-user args
+                    case ICorDebugInfo::RETBUF_ILNUM:
+                    case ICorDebugInfo::VARARGS_HND_ILNUM:
+                    case ICorDebugInfo::TYPECTXT_ILNUM:
+                    case ICorDebugInfo::UNKNOWN_ILNUM:
+                        return GCInfo::WriteBarrierForm::WBF_BarrierUnknown;
+
+                    default:
+                        break;
+                }
+                assert(ilArg != BAD_VAR_NUM);
+
+                // A small inconsistency: various signature walking VM APIs (e.g. getArgClass)
+                // do not expose implicit "this" args, while compMap2ILvarNum does.
+                if (comp->info.compMethodInfo->args.hasImplicitThis())
+                {
+                    assert(ilArg > 0);
+                    ilArg--;
+                }
+
+                // First, get corresponding CORINFO_ARG_LIST_HANDLE arg data
+                CORINFO_ARG_LIST_HANDLE currentArg = comp->info.compMethodInfo->args.args;
+                for (unsigned i = 0; i < ilArg; i++)
+                {
+                    currentArg = comp->info.compCompHnd->getArgNext(currentArg);
+                }
+
+                // Now get its type and strip `ref` so we can get CORINFO_CLASS_HANDLE of the
+                // underlying type
+                CORINFO_CLASS_HANDLE byrefCls = comp->eeGetArgClass(&comp->info.compMethodInfo->args, currentArg);
+                if (byrefCls != NO_CLASS_HANDLE)
```

Comment:
What is actually checking that this is a byref to that type? Would this potentially pass for an array as well?

Reply:
The entire code sequence is a bit unfortunate... it would be much cleaner/less scary looking if we had just saved enough information in `LclVarDsc` (or a side table, similar to `lvaParameterPassingInfo`) to get the full class handle for the parameter. I suppose the main problem around that is that we don't actually call `getArgClass` on all parameters today, only on `TYP_REF` parameters.

I wonder if we should just bite the bullet and call `getArgClass` on all parameters, or perhaps change `getArgType` to unconditionally return a class handle (today it only returns the class handle for value classes).

Reply:
Agree -- we should compute this early when we first process the method and local sigs and cache it on the local var.